docusaurus/docs/React/custom-code-examples/video-integration-100ms.mdx

```diff
@@ -0,0 +1,881 @@
+
+We’ll first create a new folder for all context objects (although we only have one, we want to have a clean structure) called `contexts` and create a file inside named `VideoContext.tsx`.
+
+Before implementing the functionality, we have to think about the functionality that we need. Our channel can have 5 different states (it can have more, but we want to keep it simple):
+
+1. There’s no call going on right now
+2. There is a call happening but the current user is not participating
+3. The user is connecting to the call
+4. The user is disconnecting from the call
+5. The user is in the call
+
+We can use an `enum` to model these different states, so let’s add this code to the `VideoContext.tsx` file:
+
+```tsx
+export enum ConnectionState {
+  NoCall,
+  CallAvailable,
+  Connecting,
+  Disconnecting,
+  InCall,
+}
+```
+
+Aside from a state variable (that is of type `ConnectionState`), we need to have multiple functions that our `VideoContext` exposes. These are:
+
+1. Creating a call
+2. Joining a call
+3. Leaving a call
+4. Ending a call
+
+None of these require parameters as we can handle all functionality inside of the `VideoContext` object itself. So let’s define an `interface` for the state of our `VideoContext` object (below the definition of the `ConnectionState`:
+
+```tsx
+interface VideoState {
+  connectionState: ConnectionState;
+  createCall: () => void;
+  joinCall: () => void;
+  leaveCall: () => void;
+  endCall: () => void;
+}
+```
+
+While we’re at it, let’s define a `defaultState` (with empty function declarations) so that we can initially create the `VideoContext` object. We’ll fill it with real code in a bit. First, add this below the `VideoState` definition:
+
+```tsx
+const defaultState: VideoState = {
+  connectionState: ConnectionState.NoCall,
+  createCall: () => {},
+  joinCall: () => {},
+  leaveCall: () => {},
+  endCall: () => {},
+};
+
+export const VideoContext = createContext<VideoState>(defaultState);
+```
+
+In order to inject the `VideoContext` into our application, we will now create a `VideoContextProvider` that takes the `VideoContext.Provider` object and allows to inject `children` (which are `ReactNode` elements).
+
+For this to work, we’ll have an internal state (using React’s `useState`) that handles the `ConnectionState` inside of our provider. We also need to create a `store` object that has implementations for all the functions we defined in our `VideoState` (that will be filled in the next chapter).
+
+Here’s the code:
+
+```tsx
+export const VideoContextProvider = ({
+  children,
+}: { children: ReactNode } => {
+  const [connectionState, setConnectionState] = useState(ConnectionState.NoCall);
+
+  const store = {
+    connectionState: connectionState,
+    createCall: () => {},
+    joinCall: () => {},
+    leaveCall: () => {},
+    endCall: () => {},
+  };
```

Comment:
```suggestion
  const store = useMemo(() => ({
    connectionState,
    createCall: () => {},
    joinCall: () => {},
    leaveCall: () => {},
    endCall: () => {},
  }), [connectionState]);
```
The reference to the `store` object and to each of the call functions would change each time this provider function runs essentially triggering updates in places which rely on it. I'd go as far as pulling each of the <create/join/leave/end>call functions to their own definitions (if they rely on some "contextual value" of the provider function then define them within provider function wrapped in useCallback but if they don't then I'd pull them out of the provider completely to keep them stable) and then passing them down to the store object adding them also to the dependency array.